src/cops.rs

```diff
@@ -0,0 +1,372 @@
+        } else {
+            return Err(BadRequest::from("Missing nglAppId field in request."));
+        }
+        if let Some(app_version) = app_map["nglAppVersion"].as_str() {
+            req.app_version = app_version.to_string();
+        } else {
+            return Err(BadRequest::from("Missing nglAppVersion field in request."));
+        }
+        if let Some(ngl_version) = app_map["nglLibVersion"].as_str() {
+            req.ngl_version = ngl_version.to_string();
+        } else {
+            return Err(BadRequest::from("Missing nglLibVersion field in request."));
+        }
+        if map.get("deviceDetails").is_none() {
+            return Err(BadRequest::from("Missing deviceDetails object in request."));
+        }
+        let device_map: HashMap<String, Value> =
+            serde_json::from_value(map["deviceDetails"].clone()).unwrap_or_default();
+        if let Some(device_date) = device_map["currentDate"].as_str() {
+            req.device_date = device_date.to_string();
+        } else {
+            return Err(BadRequest::from("Missing currentDate field in request."));
+        }
+        if let Some(device_id) = device_map["deviceId"].as_str() {
+            req.device_id = device_id.to_string();
+        } else {
+            return Err(BadRequest::from("Missing deviceId field in request."));
+        }
+        if let Some(os_user_id) = device_map["osUserId"].as_str() {
+            req.os_user_id = os_user_id.to_string();
+        } else {
+            return Err(BadRequest::from("Missing osUserId field in request."));
+        }
+        if let Some(os_name) = device_map["osName"].as_str() {
+            req.os_name = os_name.to_string();
+        } else {
+            return Err(BadRequest::from("Missing osName field in request."));
+        }
+        if let Some(os_version) = device_map["osVersion"].as_str() {
+            req.os_version = os_version.to_string();
+        } else {
+            return Err(BadRequest::from("Missing osVersion field in request."));
+        }
+        if let Some(is_vdi) = device_map["enableVdiMarkerExists"].as_bool() {
+            req.is_vdi = is_vdi;
+        } else {
+            req.is_vdi = false;
+        }
+        if let Some(is_domain_user) = device_map["isOsUserAccountInDomain"].as_bool() {
+            req.is_domain_user = is_domain_user;
+        } else {
+            req.is_domain_user = false;
+        }
+        if let Some(is_virtual) = device_map["isVirtualEnvironment"].as_bool() {
+            req.is_virtual = is_virtual;
+        } else {
+            req.is_virtual = false;
+        }
+        Ok(req)
+    }
+
+    fn from_deactivation(parts: &Parts) -> Result<Request, BadRequest> {
+        let mut req = Request {
+            kind: Kind::Deactivation,
+            timestamp: current_timestamp(),
+            ..Default::default()
+        };
+        req.update_from_headers(parts)?;
+        let pairs: HashMap<String, String> = Url::parse(&parts.uri.to_string())
```

Comment:
When I test license deactivation, I get a URL parse error - `RelativeUrlWithoutBase`. The URL that we get from the request parts appears to just include the path and query string. Example:

```
/asnp/frl_connected/v1?npdId=NTU4YThhYz...mE4YmQ1YzA1&deviceId=d5a9c61202a6a0a9...1c7681714f1b&osUserId=3b831e95ca3c1802aa687c...3a70cbc&enableVdiMarkerExists=0&isVirtualEnvironment=0&isOsUserAccountInDomain=1
```

Without the host component, the `url` crate is unable to parse the query string. One option is to add the hostname from `Settings`, but we're not getting the conf object here. It seems sufficient just to put in some placeholder. Here's a quick fix I came up with and tested successfully:

```rust
        let request_url = format!("http://placeholder{}", &parts.uri.to_string());
        let pairs: HashMap<String, String> = Url::parse(&request_url)
            .expect("Bad deactivation query string")
            .query_pairs()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect();
```

Reply:
I'm glad you tested deactivation - I hadn't yet so it doesn't surprise me there are lurking bugs.  I'll take a look and probably use your fix.